cli/dcos-spark/security.go

```diff
@@ -0,0 +1,236 @@
+			"with Kerberos principal")
+	}
+
+	if args.keytabSecretPath != "" && (args.tgtSecretValue != "" || args.tgtSecretPath != "") {
+		return errors.New("keytabs and TGTs cannot be used together")
+	}
+
+	if args.tgtSecretPath != "" && args.tgtSecretValue != "" {
+		return errors.New("cannot use a TGT-by-value and a TGT-by-secret at the same time")
+	}
+	return nil
+}
+
+func GetPrincipal(args *sparkArgs) string {
+	principal, contains := args.properties["spark.yarn.principal"]
+	if contains {
+		return principal
+	} else {
+		if args.kerberosPrincipal != "" {
+			return args.kerberosPrincipal
+		} else {
+			return ""
+		}
+	}
+}
+
+func setupKerberosSecretsConfigs(args *sparkArgs) error {
+	args.properties["spark.mesos.containerizer"] = "mesos"
+	if args.keytabSecretPath != "" { // using keytab secret
+		addArgsForFileBasedSecret(args, args.keytabSecretPath, "spark.yarn.keytab")
+		return nil
+	}
+	if args.tgtSecretPath != "" { // using tgt secret
+		addArgsForFileBasedSecret(args, args.tgtSecretPath, "spark.mesos.driverEnv.KRB5CCNAME")
+		return nil
+	}
+	if args.tgtSecretValue != "" { // using secret by value
+		appendToProperty("spark.mesos.driver.secret.values", args.tgtSecretValue, args)
+		args.properties["spark.mesos.driverEnv.KRB5CCNAME"] = "tgt"
+		appendToProperty(fmt.Sprintf(SECRET_FILENAME_TEMPLATE, "driver"), "tgt.base64", args)
+		return nil
+	}
+	return errors.New(fmt.Sprintf("Unable to add Kerberos args, got args %s", args))
+}
+
+func propertyGiven(propValue string, err error) bool {
+	if propValue == "" && err != nil {
+		return false
+	} else {
+		return true
+	}
+}
+
+func addEnvvarToDriverAndExecutor(args *sparkArgs, key, value string) {
+	driverProp := fmt.Sprintf("spark.mesos.driverEnv.%s", key)
+	executorProp := fmt.Sprintf("spark.executorEnv.%s", key)
+	_, contains := args.properties[driverProp]
+	if !contains {
+		args.properties[driverProp] = value
+	}
+	_, contains = args.properties[executorProp]
+	if !contains {
+		args.properties[executorProp] = value
+	}
+}
+
+func forwardEnvironmentVariablesFromMarathonConfig(args *sparkArgs, marathonJson map[string]interface{}) {
+	krb5confBase64, errBase64Blob := getStringFromTree(marathonJson, []string{"app", "env", sparkKrb5ConfBase64BlobKey})
+	kdcHostname, errHostname := getStringFromTree(marathonJson, []string{"app", "env", sparkKdcHostnameKey})
+	kdcPort, errPort := getStringFromTree(marathonJson, []string{"app", "env", sparkKdcPortKey})
+	kdcRealm, errRealm := getStringFromTree(marathonJson, []string{"app", "env", sparkKerberosRealmKey})
+
+	// We allow the user to set SPARK_SECURITY_KERBEROS_KDC_HOSTNAME and SPARK_SECURITY_KERBEROS_KDC_PORT, and
+	// SPARK_SECURITY_KERBEROS_REALM, these values will be used to template a krb5.conf. If the user sets
+	// SPARK_MESOS_KRB5_CONF_BASE64 it will be overwritten, but log a warning to be sure.
+	kdcPropCount := 0
+	if propertyGiven(kdcHostname, errHostname) {
```

Comment:
The combination of `getStringFromTree` and `if propertyGiven()` block and be combined into a function.